xdc-plugin/xdc.cc

```diff
@@ -0,0 +1,242 @@
+		log("\n");
+		log("    set_property PROPERTY VALUE OBJECT\n");
+		log("\n");
+		log("Set the given property to the specified value on an object\n");
+		log("\n");
+	}
+
+	void execute(std::vector<std::string> args, RTLIL::Design* design) YS_OVERRIDE {
+		if (design->top_module() == nullptr) {
+			log_cmd_error("No top module detected\n");
+		}
+
+		std::string option(args[1]);
+		if (set_property_options_map.count(option) == 0) {
+			log_warning("set_property: %s option is currently not supported\n", option.c_str());
+			return;
+		}
+
+		switch (set_property_options_map[option]) {
+			case INTERNAL_VREF:
+				process_vref(std::vector<std::string>(args.begin() + 2, args.end()), design);
+				break;
+			default:
+				assert(false);
+		}
+	}
+
+	void process_vref(std::vector<std::string> args, RTLIL::Design* design) {
+		if (args.size() < 2) {
+			log_error("set_property INTERNAL_VREF: Incorrect number of arguments.\n");
+		}
+		int iobank = std::atoi(args[1].c_str());
+		if (bank_tiles.count(iobank) == 0) {
+			log_cmd_error("set_property INTERNAL_VREF: Invalid IO bank.\n");
+		}
+
+		int internal_vref = 1000 * std::atof(args[0].c_str());
+		if (internal_vref != 600 &&
+				internal_vref != 675 &&
+				internal_vref != 750 &&
+				internal_vref != 900) {
+			log("set_property INTERNAL_VREF: Incorrect INTERNAL_VREF value\n");
+			return;
+		}
+
+		// Create a new BANK module if it hasn't been created so far
+		RTLIL::Module* top_module = design->top_module();
+		if (!design->has(ID(BANK))) {
+			std::string fasm_extra_modules_dir(proc_share_dirname() + "/plugins/fasm_extra_modules");
+			Pass::call(design, "read_verilog " + fasm_extra_modules_dir + "/BANK.v");
+		}
+
+		// Set parameters on a new bank instance or update an existing one
+		char bank_cell_name[16];
+		snprintf(bank_cell_name, 16, "\\bank_cell_%d", iobank);
+		RTLIL::Cell* bank_cell = top_module->cell(RTLIL::IdString(bank_cell_name));
+		if (!bank_cell) {
+			bank_cell = top_module->addCell(RTLIL::IdString(bank_cell_name), ID(BANK));
+		}
+		bank_cell->setParam(ID(FASM_EXTRA), RTLIL::Const("INTERNAL_VREF"));
+		bank_cell->setParam(ID(NUMBER), RTLIL::Const(iobank));
+		bank_cell->setParam(ID(INTERNAL_VREF), RTLIL::Const(internal_vref));
+	}
+
+} SetProperty;
+
+struct GetBankTiles : public Pass {
+	GetBankTiles() : Pass("get_bank_tiles", "Inspect IO Bank tiles") {
+		register_in_tcl_interpreter(pass_name);
+	}
+
+	void help() YS_OVERRIDE
+	{
+		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
+		log("\n");
+		log("   get_bank_tiles <part_json_file>\n");
+		log("\n");
+		log("Inspect IO Bank tiles for the specified part based on the provided JSON file.\n");
+		log("\n");
+	}
+
+	void execute(std::vector<std::string> args, RTLIL::Design* ) YS_OVERRIDE {
+                if (args.size() < 2) {
+                        log_cmd_error("Missing JSON file.\n");
+		}
+		// Check if the part has the specified bank
```

Comment:
There is not reason to use the global variable bank_tiles here.

Reply:
Sure, fixed.